src/Component/tests/Doctrine/Persistence/InMemoryRepositoryTest.php

```diff
@@ -0,0 +1,266 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Resource\Tests\Doctrine\Persistence;
+
+use Pagerfanta\Pagerfanta;
+use PHPUnit\Framework\TestCase;
+use Prophecy\PhpUnit\ProphecyTrait;
+use Prophecy\Prophecy\ObjectProphecy;
+use Sylius\Resource\Doctrine\Persistence\Exception\ResourceExistsException;
+use Sylius\Resource\Doctrine\Persistence\InMemoryRepository;
+use Sylius\Resource\Doctrine\Persistence\RepositoryInterface;
+use Sylius\Resource\Exception\UnexpectedTypeException;
+use Sylius\Resource\Model\ResourceInterface;
+
+final class InMemoryRepositoryTest extends TestCase
+{
+    use ProphecyTrait;
+
+    private InMemoryRepository $repository;
+
+    protected function setUp(): void
+    {
+        $this->repository = new InMemoryRepository(SampleBookResourceInterface::class);
+    }
+
+    public function testItThrowsUnexpectedTypeExceptionWhenConstructingWithoutResourceInterface(): void
+    {
+        $this->expectException(UnexpectedTypeException::class);
+        new InMemoryRepository(\stdClass::class);
+    }
+
+    public function testItImplementsRepositoryInterface(): void
+    {
+        $this->assertInstanceOf(RepositoryInterface::class, $this->repository);
+    }
+
+    public function testItThrowsInvalidArgumentExceptionWhenAddingWrongResourceType(): void
+    {
+        /** @var ObjectProphecy<ResourceInterface> $resource */
+        $resource = $this->prophesize(ResourceInterface::class);
+
+        $this->expectException(\InvalidArgumentException::class);
```

Comment:
Outside of scope of current PR, but there should be specialized exception (Not global `\InvalidArgumentException`) thrown by the repository.